src/main/java/com/commercetools/pspadapter/notification/processor/impl/NotificationProcessorBase.java

```diff
@@ -0,0 +1,97 @@
+import io.sphere.sdk.payments.commands.updateactions.ChangeTransactionState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.time.ZonedDateTime;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+import static java.lang.String.format;
+
+public abstract class NotificationProcessorBase implements NotificationProcessor {
+
+    private static final Logger logger = LoggerFactory.getLogger(NotificationProcessorBase.class);
+
+    private final Gson gson;
+
+    NotificationProcessorBase(Gson gson) {
+        this.gson = gson;
+    }
+
+    abstract Optional<ChangeTransactionState> createChangeTransactionState(@Nonnull Payment ctpPayment);
+
+    @Override
+    public CompletionStage<Payment> processEventNotification(@Nonnull CtpFacade ctpFacade,
+                                                             @Nonnull Event event) {
+        if (!canProcess(event)) {
+            throw new IllegalArgumentException(format("Event can't be processed. [%s]", event.toJSON()));
+        }
+        return getRelatedCtpPayment(ctpFacade, event)
+                .thenCompose(ctpPaymentOpt -> ctpPaymentOpt.map(ctpPayment -> ctpFacade.getPaymentService()
+                        .updatePayment(ctpPayment, createPaymentUpdates(ctpPayment, event)))
+                        .orElseThrow(() -> new NotFoundException(format("No related CTP payment found for event [%s]", event.toJSON())))
+                );
+    }
+
+    protected ImmutableList<UpdateAction<Payment>> createPaymentUpdates(@Nonnull Payment ctpPayment,
+                                                                        @Nonnull Event event) {
+        final ImmutableList.Builder<UpdateAction<Payment>> listBuilder = ImmutableList.builder();
+        listBuilder.add(createAddInterfaceInteractionAction(event));
+        Optional<ChangeTransactionState> changeTransactionOpt = createChangeTransactionState(ctpPayment);
+        if (changeTransactionOpt.isPresent()) {
+            listBuilder.add(changeTransactionOpt.get());
+        } else {
+            logger.warn("Notification event {} did not trigger change transaction state", event);
+        }
+        return listBuilder.build();
+    }
+
+    @SuppressWarnings("unchecked")
+    protected CompletionStage<Optional<Payment>> getRelatedCtpPayment(@Nonnull CtpFacade ctpFacade,
+                                                                      @Nonnull Event event) {
+        Map<String, String> resource = (Map<String, String>) event.getResource();
+        String ppPlusPaymentId = resource.get("parent_payment");
+
+        return ctpFacade.getPaymentService()
+                .getByPaymentInterfaceNameAndInterfaceId(PaypalPlusPaymentInterfaceName.PAYPAL_PLUS, ppPlusPaymentId);
+    }
+
+    protected Optional<Transaction> findMatchingTxn(@Nonnull Collection<Transaction> transactions,
```

Comment:
looks like static util method

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/57/commits/77fbacbb6de52feef0c0d05225c6ecf771b94d84

Reply:
i didn't mean public util method (i just mentioned that method could be static in the same class), but anyway - looks better!